File: sem3/map/ToyLanguageInterpreter/src/model/statements/ReadFileStatement.java
package model.statements;

import model.ProgramState;
import model.expressions.ConstExpr;
import model.expressions.Expression;
import utils.FileData;
import utils.FileTable;
import utils.exceptions.InterpreterException;

import java.io.BufferedReader;
import java.io.IOException;

/**
 * Created by mirko on 09/11/2016.
 */
public class ReadFileStatement implements Statement {
    private Expression fileDescriptor;
    private String variable;

    public ReadFileStatement(Expression fileDescriptor, String variable) {
        this.fileDescriptor = fileDescriptor;
        this.variable = variable;
    }

    @Override
    public ProgramState execute(ProgramState programState) throws InterpreterException {
        try {
            int descriptor = fileDescriptor.evaluate(programState.getSymbolTable(), programState.getHeap());
            FileTable<Integer, FileData<String, BufferedReader>> fileTable = programState.getFileTable();
            BufferedReader bufferedReader = fileTable.getValue(descriptor).getValue();

            String line = bufferedReader.readLine();
            ConstExpr value;

            if (line == null) {
                value = new ConstExpr(0);
            } else {
                value = new ConstExpr(Integer.parseInt(line));
            }

            Statement assignment = new AssignmentStatement(variable, value);
            return assignment.execute(programState);

        } catch (IOException e) {
            throw new InterpreterException("error: file could not be read");
        }
    }

    @Override
    public String toString() {
        return "readFile(" + fileDescriptor + ", " + variable + ")";
    }
}

File: sem3/map/ToySharpLanguageInterpreter/ToySharpLanguageInterpreter/src/model/statements/ReadFileStatement.cs
using System;
using System.IO;

namespace ToySharpLanguageInterpreter
{
	public class ReadFileStatement : Statement
	{
		private Expression fileDescriptor;
		private string variable;

		public ReadFileStatement(Expression fileDescriptor, string variable)
		{
			this.fileDescriptor = fileDescriptor;
			this.variable = variable;
		}


		public ProgramState execute(ProgramState programState)
		{
			try
			{
				int descriptor = fileDescriptor.evaluate(programState.getSymbolTable());
				FileTable<int, FileData<string, StreamReader>> fileTable = programState.getFileTable();
				StreamReader bufferedReader = fileTable.getValue(descriptor).Item2;

				string line = bufferedReader.ReadLine();
				ConstExpr value;

				if (line == null)
				{
					value = new ConstExpr(0);
				}
				else {
					value = new ConstExpr(int.Parse(line));
				}

				Statement assignment = new AssignmentStatement(variable, value);

				return assignment.execute(programState);

			}
			catch (IOException)
			{
				throw (new InterpreterException("error: file could not be read"));
			}
		}


		public override string ToString()
		{
			return "readFile(" + fileDescriptor + ", " + variable + ")";
		}
	}
}